projects/rolling-statistics/rolling_statistics.py

```diff
@@ -0,0 +1,699 @@
+    five_min_ts = pd.date_range(first_date, last_date, freq="5min")
+    new_df = pd.DataFrame({"est.localTime_rounded": five_min_ts})
+
+    # round est.localTime to nearest 30 seconds, then 5 minutes
+
+    # SLOW BUT MORE ACCURATE
+    # df = round5Minutes(df)
+
+    # FAST BUT LESS ACCURATE
+    df["est.localTime_rounded"] = pd.to_datetime(pd.to_datetime(df["est.localTime"]).astype(np.int64)+1000).dt.round("30S")
+    df["est.localTime_rounded"] = pd.to_datetime(pd.to_datetime(df["est.localTime_rounded"]).astype(np.int64)+1000).dt.round("5min")
+
+    print("done")
+
+    df, rounded_duplicates = remove_rounded_duplicates(df.copy(), data_type)
+    new_df = pd.merge(new_df, df, on="est.localTime_rounded", how="outer", indicator=True)
+
+    # If basal data, forward fill the rates by duration
+    if data_type == "basal":
+        fill_basal_gaps(new_df)
+
+    return new_df, rounded_duplicates
+
+
+def get_rolling_stats(df, rolling_prefixes):
+
+    print("Preparing Variables... ", end=" ")
+
+    # Functions to calculate average daily hypo/hyper events and duration
+    def rle(inarray):
+        """ run length encoding. Partial credit to R rle function.
+            Multi datatype arrays catered for including non Numpy
+            returns: tuple (runlengths, startpositions, values) """
+        ia = np.asarray(inarray)                  # force numpy
+        n = len(ia)
+        if n == 0:
+            return (None, None, None)
+        else:
+            y = np.array(ia[1:] != ia[:-1])      # pairwise unequal (string safe)
+            i = np.append(np.where(y), n - 1)    # must include last element posi
+            z = np.diff(np.append(-1, i))        # run lengths
+            p = np.cumsum(np.append(0, z))[:-1]  # positions
+            return(z, p, ia[i])
+
+    # Setup run length encoding for hypo/hyper events
+    rle_below54 = rle(df.mg_dL < 54)
+    rle_below70 = rle(df.mg_dL < 70)
+    rle_above140 = rle(df.mg_dL > 140)
+    rle_above180 = rle(df.mg_dL > 180)
+    rle_above250 = rle(df.mg_dL > 250)
+
+    col_names = ["event-below54",
+                 "dur-below54",
+                 "event-below70",
+                 "dur-below70",
+                 "event-above140",
+                 "dur-above140",
+                 "event-above180",
+                 "dur-above180",
+                 "event-above250",
+                 "dur-above250"]
+
+    df.reindex(columns=[df.columns.tolist()+col_names])
+    # Start getting locations and durations of hypo/hyper episodes
+    # Note: Error 712 is common here due to bool checks in a numpy function
+    # See https://github.com/PyCQA/pycodestyle/issues/450
+
+    below54_loc = rle_below54[1][np.where((rle_below54[2] == True) & (rle_below54[0] >= 3))]
+    below54_dur = 5*rle_below54[0][np.where((rle_below54[2] == True) & (rle_below54[0] >= 3))]
+    df["event-below54"] = False
+    df.loc[below54_loc, "event-below54"] = True
+    df["dur-below54"] = 0
```

Comment:
when specifying durations, I find that including the units in the variable name or colHeading can be useful. Good to know that we are talking about minutes here.